text/0000-natvis.md

```diff
@@ -0,0 +1,491 @@
+
+This section describes how Microsoft's Natvis is supported in Rust.
+
+To use Natvis, developers write XML documents that describe how debugger types
+should be displayed using the natvis schema. (See: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019)
+The Natvis files provide patterns, which match type names, and for matching
+types, a description of how to display those types. This allows for some
+limited support for generic types.
+
+Rust developers can add one or more `.natvis` files to their crate. The
+`Cargo.toml` file specifies the path of these Natvis files via a new
+manifest key. Cargo would then pass the set of `.natvis` files that were
+specified in `Cargo.toml` and pass it to the rustc invocation for the crate
+via a new `-C` flag and enabling unstable options in Rust via the
+`-Z unstable-options`. See the below example for how rustc would embed these
+Natvis files in the debuginfo for a binary as well as new compiler options
+to be added to rustc.
+
+To provide Natvis files, developers create a file with the `.natvis` file
+extension and specify this Natvis file in the `Cargo.toml` under a new section
+that will be added as part of this RFC.
+
+As an example, consider a crate `foo` with this directory structure:
+
+```text
+/Cargo.toml
+/Foo.natvis (Note: the .natvis file does not have to match the name of the crate.)
+  +-- src
+      +-- main.rs
+```
+
+Where `main.rs` contains:
+
+```rust
+/// A rectangle in first quadrant
+struct FancyRect {
+    pub x: f32,
+    pub y: f32,
+    pub dx: f32,
+    pub dy: f32,
+}
+
+fn main() {
+    let mut fancy_rect = FancyRect::new(10.0, 10.0, 5.0, 5.0);
+    println!("FancyRect: {:?}", fancy_rect);
+}
+```
+
+and `Foo.natvis` contains:
+
+```xml
+<?xml version="1.0" encoding="utf-8"?>
+<AutoVisualizer xmlns="http://schemas.microsoft.com/vstudio/debugger/natvis/2010">
+    <Type Name="foo::FancyRect">
+      <DisplayString>({x},{y}) + ({dx}, {dy})</DisplayString>
+      <Expand>
+        <Synthetic Name="LowerLeft">
+          <DisplayString>({x}, {y})</DisplayString>
+        </Synthetic>
+        <Synthetic Name="UpperLeft">
+          <DisplayString>({x}, {y + dy})</DisplayString>
+        </Synthetic>
+        <Synthetic Name="UpperRight">
+          <DisplayString>({x + dx}, {y + dy})</DisplayString>
+        </Synthetic>
+        <Synthetic Name="LowerRight">
+          <DisplayString>({x + dx}, {y})</DisplayString>
+        </Synthetic>
+      </Expand>
+    </Type>
+</AutoVisualizer>
+```
+
+When viewed under WinDbg, the `fancy_rect` variable would be shown as follows:
+
+```text
+> Variables:
+  > fancy_rect: (10, 10) + (5, 5)
+    > LowerLeft: (10, 10)
+    > UpperLeft: (10, 15)
+    > UpperRight: (15, 15)
+    > LowerRight: (15, 10)
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In rustc, a new `-C natvis={comma-separated list of .natvis files}` flag
```

Comment:
```suggestion
In rustc, a new `-C natvis={comma-separated list of paths}` flag
```

Reply:
I would prefer to keep the `.natvis` extension here. We could clarify that the `.natvis` extension is only a convention, and that `rustc` does not require that these files have a specific extension, but it does help keep this explanatory text clear that the `-C natvis` argument value points to a specific kind of file, and not a directory.

Reply:
I agree, keeping this text in the documentation will clarify exactly what kind of files are expected here. 

> We could clarify that the .natvis extension is only a convention, and that rustc does not require that these files have a specific extension

@sivadeilra is there any reason we do not want to require these files to have the `.natvis` extension in rustc? WinDbg is unable to load Natvis files embedded into the PDB if they do not carry the `.natvis` extension but the linker will not emit any errors and still embed this file into the PDB.

Reply:
Ah, I wasn't aware that WinDbg would not load the NatVis stream if the original file did not have a `.natvis` extension.

How about this:

1.  `rustc` doesn't care what the file extension is -- it will do what it's told, and pass through any file as a NatVis stream to the linker.
2. Cargo will only discover NatVis files that _do_ have the `.natvis` extension, in the appropriate debugger visualizations subdirectory.
3. If you use the explicit manifest key for including NatVis streams, then Cargo will check the file extension for any file explicitly specified in the manifest.  If the extension is not `.natvis`, then Cargo will print a warning, but still pass the `-C natvis` argument.

I think this will give a good development experience, but it also avoids baking in too much knowledge about what the file extension is, to rustc itself.  I'm guessing that the `.natvis` file extension requirement in WinDbg is an implementation limitation, and not an intentional choice.

Reply:
> I think this will give a good development experience, but it also avoids baking in too much knowledge about what the file extension is, to rustc itself.

This sounds good to me. I'll make note of the Cargo warning in the RFC.